src/pyspiffe/bundle/x509_bundle/x509_bundle_set.py

```diff
@@ -13,26 +14,24 @@
 class X509BundleSet(object):
     """X509BundleSet is a set of X509Bundles objects, keyed by trust domain."""
 
-    def __init__(self, bundles_map: Mapping[TrustDomain, X509Bundle]) -> None:
+    def __init__(self, bundles_map: Dict[TrustDomain, X509Bundle]) -> None:
         """Creates a new initialized X509BundleSet with the given X509Bundle objects keyed by TrustDomain.
 
         Args:
             bundles_map: A map object of X509Bundle objects keyed by TrustDomain to initialize the X509BundleSet.
         """
 
-        # create and initialize a dict, so it can be updated (as Mapping doesn't define __setitem__)
-        bundles = {}
-        for td, bundle in bundles_map.items():
-            bundles[td] = bundle
-        self._bundles = bundles
+        self.lock = threading.Lock()
+        self._bundles = copy.copy(bundles_map)
 
     def put(self, bundle: X509Bundle) -> None:
         """Adds a new X509Bundle object or replace an existing one into the set.
 
         Args:
             bundle: The new X509Bundle to put into the set.
         """
-        self._bundles[bundle.trust_domain()] = bundle
+        with self.lock:
+            self._bundles[bundle.trust_domain()] = bundle
```

Comment:
should this method store a copy of the input bundle object?

Reply:
As the bundle itself "protects" its internal collections through safe access methods, I think it is safe to share bundles and this way we avoid incurring in overheads needing to deep copy everything in each put/get method.